astropy/coordinates/baseframe.py

```diff
@@ -333,39 +335,31 @@ def __init__(
                 f"keywords: {list(kwargs)}"
             )
 
-        # We do ``is None`` because self._data might evaluate to false for
-        # empty arrays or data == 0
-        if self._data is None:
-            # No data: we still need to check that any non-scalar attributes
-            # have consistent shapes. Collect them for all attributes with
-            # size > 1 (which should be array-like and thus have a shape).
-            shapes = {
-                fnm: value.shape
-                for fnm, value in values.items()
-                if getattr(value, "shape", ())
-            }
-            if shapes:
-                if len(shapes) > 1:
-                    try:
-                        self._no_data_shape = check_broadcast(*shapes.values())
-                    except ValueError as err:
-                        raise ValueError(
-                            f"non-scalar attributes with inconsistent shapes: {shapes}"
-                        ) from err
-
-                    # Above, we checked that it is possible to broadcast all
-                    # shapes.  By getting and thus validating the attributes,
-                    # we verify that the attributes can in fact be broadcast.
-                    for fnm in shapes:
-                        getattr(self, fnm)
+        # Determine the overall shape of the frame.
+        try:
+            self._shape = check_broadcast(*shapes)
+        except ValueError as err:
+            raise ValueError(
+                f"non-scalar data and/or attributes with inconsistent shapes: {shapes}"
+            ) from err
+
+        # Broadcast the data if necessary and set it
+        if data is not None and data.shape != self._shape:
+            data = data._apply(np.broadcast_to, shape=self._shape, subok=True)
+        self._data = data
+
+        # Broadcast and set the attributes (leave scalars as is)
+        for key, value in values.items():
+            shape = getattr(value, "shape", ())
```

Comment:
See my explanation https://github.com/astropy/astropy/pull/15121#discussion_r1342873586. We can't broadcast attributes here because we don't know the overall shape of the frame yet.